File: day5_part1_2017.exs

defmodule MazeSolver do
  def call do
    "input.txt"
    |> File.read!()
    |> String.split("\n", trim: true)
    |> Enum.map(&String.to_integer/1)
    |> solve(0, 0)
  end

  defp solve(instructions, position, steps) when position < 0 or position >= length(instructions), do: steps
  defp solve(instructions, position, steps) do
    jump = Enum.at(instructions, position)
    updated_instructions = List.replace_at(instructions, position, jump + 1)
    solve(updated_instructions, position + jump, steps + 1)
  end
end

File: training_data/reflection-tuning/day5_part1_2017.clj
(defn solve-maze [instructions]
  (loop [pos 0
         steps 0
         maze instructions]
    (if (or (< pos 0) (>= pos (count maze)))
      steps
      (let [jump (get maze pos)]
        (recur (+ pos jump)
               (inc steps)
               (assoc maze pos (inc jump)))))))

(defn read-input [file]
  (vec (map #(Integer/parseInt %) (clojure.string/split-lines (slurp file)))))

(defn -main []
  (let [input (read-input "input.txt")
        result (solve-maze input)]
    (println result)))

(-main)
